[aes_gcm,sca] Add tests

nasahlpa opened the pull request:
This commit extends the uJSON framework with the following two tests:

- Single AES-GCM operation. The device performs a single AES-GCM operation on a given input. The host can configure the number of AAD and PTX blocks as well as when the trigger pin gets asserted.
- Batch FvsR AES-GCM operation. A fixed vs. random set is executed N(=number of batch operations) times

nasahlpa committed:
[aes_gcm,sca] Add single encrypt

This commit extends the uJSON framework with the possibility
of running a single AES GCM operation. The host needs to send
all required input, the ciphertext and the tag is returned to
the host.

Signed-off-by: Pascal Nasahl <nasahlpa@lowrisc.org>

nasahlpa commented:
Tested on CW310 with the bitstream build from the a`es-gcm-review` branch. In the `kDifAesManualOperationAuto` without the `aes_manual_trigger` commands, the Python AES GCM library we use in ot-sca & this code produces the correct tag & ciphertext.

vogelpi commented on sw/device/tests/penetrationtests/firmware/sca/aes_sca.c:
```diff
@@ -630,6 +630,129 @@ status_t handle_aes_sca_fvsr_key_start_batch_generate(ujson_t *uj) {
+  TRY(dif_aes_load_data(&aes, ptx));
+  aes_manual_trigger();
+
+  // Read the CTX block from the AES.
+  AES_TESTUTILS_WAIT_FOR_STATUS(&aes, kDifAesStatusOutputValid, true,
+                                1000 * 1000);
+  TRY(dif_aes_read_output(&aes, ctx));
```
Nit: you don't actually need to read the ciphertext. The tag is sufficient to verify correct operation :-)

nasahlpa replied:
Makes sense to remove it once everything works (this allows us to get more traces/s). But for debugging I leave it for now.

nasahlpa committed:
[aes,sca] Add AES-GCM FvsR batch test

This commit adds the AES-GCM fixed vs random batch test.
The host configures:
- Number of batch operation s(max 256)
- Where to trigger
- The fixed IV & fixed key
- The number of AAD and PTX blocks
- The fixed AAD and PTX blocks

The FvsR set consists of the IV and the key, i.e.,
on the device a fixed and a random set of these variables
are created.

The AES-GCM operation is executed batch_operations. The XOR
of all tags / ctxs are transmitted back to the host for
verification.

Signed-off-by: Pascal Nasahl <nasahlpa@lowrisc.org>

nasahlpa commented:
Ready for review @vogelpi & @andrea-caforio. Tested on CW310.

I'll remove the CTX after the review with an additional commit - if I need to change something having the CTX makes it easier to debug :-)

vogelpi commented on sw/device/tests/penetrationtests/firmware/sca/aes_sca.c:
```diff
@@ -630,6 +643,409 @@ status_t handle_aes_sca_fvsr_key_start_batch_generate(ujson_t *uj) {
+  AES_TESTUTILS_WAIT_FOR_STATUS(&aes, kDifAesStatusInputReady, true,
+                                kIbexAesGcmSleepCycles * 2);
+  TRY(dif_aes_load_gcm_tag_len(&aes, len_ptx, len_aad));
+  aes_manual_trigger();
```
Nit: this tag block would also be an analysis target. It's certain lower priority but it could be interesting to see that as well.

nasahlpa replied:
Done, thanks for the feedback.

vogelpi commented on sw/device/tests/penetrationtests/firmware/sca/aes_sca.c:
```diff
@@ -630,6 +643,409 @@ status_t handle_aes_sca_fvsr_key_start_batch_generate(ujson_t *uj) {
+      uj_ptx_blocks.num_blocks > kMaxGcmBlocks) {
+    return ABORTED();
+  }
+  // Fetch all AAD blocks.
+  for (size_t it = 0; it < uj_aad_blocks.num_blocks; it++) {
+    TRY(ujson_deserialize_aes_sca_block_t(uj, &uj_aad[it]));
+  }
+  // Fetch all PTX blocks.
+  for (size_t it = 0; it < uj_ptx_blocks.num_blocks; it++) {
+    TRY(ujson_deserialize_aes_sca_block_t(uj, &uj_ptx[it]));
+  }
```
I was first surprised to see that we're getting all AAD and PTX blocks but it makes sense as we keep those fixed: we actually want to analyze the key and IV.

vogelpi reviewed (approved):
Thanks @nasahlpa , this looks fantastic! Great work!

The only thing I would potentially add is the possibility to analyze the tag generation. Because this one performs a series of steps (including the unmasking of the state and adding the last correction term) which means exercising some additional datapaths not exercised during the other blocks.

nasahlpa committed:
[aes,sca] Add trigger to tag generation

This commit extends the aes_gcm function such that the user
can also capture the tag generation.

Signed-off-by: Pascal Nasahl <nasahlpa@lowrisc.org>

nasahlpa committed:
[aes,sca] Remove CTX from SCA tests

As the tag is sufficient to check whether the AES-GCM operation
was correct, this commit removes reading out and sending back
the CTX to the host. This allows us to achieve faster capture
rates.

Signed-off-by: Pascal Nasahl <nasahlpa@lowrisc.org>

nasahlpa committed:
[aes_gcm,sca] Adapt sleep cycles

This commit adapts the number of cycles we are putting Ibex into
sleep. The previous cycle count was too large as for the mcycle
measurement the AES_TESTUTILS_WAIT_FOR_STATUS function was used
to wait for the IDLE bit. By directly using the DIF to wait for
this bit, we are reducing the overhead caused by this library.

Signed-off-by: Pascal Nasahl <nasahlpa@lowrisc.org>

nasahlpa committed:
[aes_gcm,sca] Fix 0 init of array

This commit fixes the zero init of the accumulation array.

Signed-off-by: Pascal Nasahl <nasahlpa@lowrisc.org>

nasahlpa committed:
[aes_gcm,sca] Make force_masks configurable

To enable/disable masking for AES-GCM, this commit uses the force_masks
configuration from transaction. When sending the 0 seed to the AES
application, the masking is turned off.

Signed-off-by: Pascal Nasahl <nasahlpa@lowrisc.org>

nasahlpa committed:
[aes,sca] Fix masks_off

With the recent changes in the PRNG, it is necessary to reseed the
internal AES state after the AUX register has been set by using the
dif_aes_start() function. We need to do this when we want to switch
off masking for AES SCA tests.

Signed-off-by: Pascal Nasahl <nasahlpa@lowrisc.org>

nasahlpa commented:
@vogelpi I've added some more commits after your review. As this is the code that I have used for the evaluation, can I merge it or do you want to have another look?

nasahlpa committed:
[aes,sca] Allow capturing the all-zero block for AES-GCM

This commit adds a new trigger option allowing ot-sca to also
capture the processing of the all-zero block.

Signed-off-by: Pascal Nasahl <nasahlpa@lowrisc.org>

vogelpi commented:
Thanks Pascal, this sounds good to me!
